### PR TITLE
Generic lev comp

### PR DESCRIPTION
**Pull request by jmonteer**
- Adapted WBTC Lev Comp to be able to handle other tokens.
- Removed sell of comp during withdrawal
- Added the toggle between UniV2, SushiV2 and UniV3 to dispose comp

### CODE REVIEW

File: contracts/LevCompFactory.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    address public immutable original;
+
+    event Cloned(address indexed clone);
+    event Deployed(address indexed original);
+
+    constructor(
+        address _vault,
+        address _cToken
+    ) public {
+        Strategy _original = new Strategy(_vault, _cToken);
```
Comment (storming0x) [inline contracts/LevCompFactory.sol:17]: why does this needs to wrap the Strategy API?

We are just storing the address not sure is needed.

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -39,66 +24,119 @@ interface IUni{
...
+    function approveTokenMax(address token, address spender) internal {
+        IERC20(token).safeApprove(spender, type(uint256).max);
+    }
 
+    // To receive ETH from compound and WETH contract
+    receive() external payable {}
 
-    constructor(address _vault, address _cToken) public BaseStrategy(_vault) {
+    function name() external override view returns (string memory){
+        return "StrategyGenericLevCompFarm";
```
Comment (storming0x) [inline contracts/Strategy.sol]: i think the change is big enough to add a V2 at the end of this, so we can make a new risk group for this contract and score it separately from original

File: contracts/LevCompFactory.sol [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        emit Deployed(address(_original));
+
+        original = address(_original);
+    }
+
+    function name() external pure returns (string memory) {
```
Comment (storming0x) [inline contracts/LevCompFactory.sol]: if we have several factories, adding the apiVersion would help too before cloning?

File: contracts/WBTCFlashLoan.sol [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        require(wbtc.balanceOf(address(this)) == 0);
+        require(address(this).balance == 0);
+    }
+
+    function callFunction(
+        address sender,
+        Account.Info memory accountInfo,
+        bytes memory data
+    ) external {
+        require(sender == address(this));
```
Comment (storming0x) [inline contracts/WBTCFlashLoan.sol]: Critical issue, requires msg.sender to be SOLO

File: contracts/Strategy.sol (hunk 2 above)
Comment (jmonteer) [reply contracts/Strategy.sol]: agree, changed to `GenLevCompV2`

File: contracts/LevCompFactory.sol (hunk 3 above)
Comment (jmonteer) [reply contracts/LevCompFactory.sol]: good point, adding it

Comment (jmonteer) [reply contracts/LevCompFactory.sol]: I see it was statically added. I will concat the original's version here

File: contracts/WBTCFlashLoan.sol (hunk 4 above)
Comment (jmonteer) [reply contracts/WBTCFlashLoan.sol]: this contract was not supposed to be here. deleted

thanks

File: contracts/FlashLoanLib.sol [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		Account.Info[] memory accountInfos = new Account.Info[](1);
+		accountInfos[0] = _getAccountInfo();
+
+		SOLO.operate(accountInfos, operations);
+
+		emit Leverage(amountDesired, requiredETH, deficit, address(SOLO));
+
+		return amountWant; // we need to return the amount of Want we have changed our position in
+	}
+	event Number(string name, uint number);
```
Comment (fp-crypto) [inline contracts/FlashLoanLib.sol]: Is this for debugging?

Comment (jmonteer) [reply contracts/FlashLoanLib.sol]: it is, removed

**Pull request merged**